File: cpp/factory_method/factory_method.cpp
#include <iostream>
#include <memory>

class Pizza {
public:
    virtual void bake() = 0;
};

class CheesePizza : public Pizza {
public:
    void bake() override {
        std::cout << "Baking cheese pizza!" << std::endl;
    }
};

enum PizzaType {
    Cheese
};

class PizzaStore {
public:
    std::shared_ptr<Pizza> order_pizza(PizzaType type) {
        auto pizza = create_pizza(type);
        pizza->bake();
        return pizza;
    }

protected:
    virtual std::shared_ptr<Pizza> create_pizza(PizzaType type) = 0;
};

class NYPizzaStore : public PizzaStore {
protected:
    std::shared_ptr<Pizza> create_pizza(PizzaType type) override {
        if (type == Cheese) {
            return std::make_shared<CheesePizza>();
        }
        throw std::invalid_argument("Incorrect pizza type");
    }
};

int main() {
    NYPizzaStore store;
    store.order_pizza(Cheese);
    return 0;
}

File: go/factory_method/factory_method.go
package main

import "fmt"

type Pizza struct {
}

func (p Pizza) bake() {
}

type PizzaType string

const (
	Cheese PizzaType = "Cheese"
)

type CheesePizza struct {
	Pizza
}

func (p CheesePizza) bake() {
	fmt.Println("Baking cheese pizza!")
}

type PizzaStore struct {
}

func (s PizzaStore) createPizza(t PizzaType) Pizza {
	
}

func (s PizzaStore) orderPizza()  {
}

func main() {

}
